PSReadLine/Render.cs

```diff
@@ -340,135 +347,330 @@ void MaybeEmphasize(int i, string currColor)
+        {
+            int cnt = 1;
+            int bufferWidth = _console.BufferWidth;
 
-            // Possibly need to flip the color on the prompt if the error state changed.
-            var promptText = _options.PromptText;
-            if (!string.IsNullOrEmpty(promptText))
+            if (isFirstLogicalLine)
             {
-                renderData.errorPrompt = (_parseErrors != null && _parseErrors.Length > 0);
-                if (renderData.errorPrompt != _previousRender.errorPrompt)
+                // The first logical line has the user prompt that we don't touch
+                // (except where we turn part to red, but we've finished that
+                // before getting here.)
+                var maxFirstLine = bufferWidth - _initialX;
+                if (columns > maxFirstLine)
+                {
+                    cnt += 1;
+                    columns -= maxFirstLine;
+                }
+                else
                 {
-                    // We need to update the prompt
+                    lenLastPhysicalLine = columns;
+                    return 1;
+                }
+            }
+
+            lenLastPhysicalLine = columns % bufferWidth;
+            if (lenLastPhysicalLine == 0)
+            {
+                // Handle the last column when the columns is equal to n * bufferWidth
+                // where n >= 1 integers
+                lenLastPhysicalLine = bufferWidth;
+                return cnt - 1 + columns / bufferWidth;
+            }
+
+            return cnt + columns / bufferWidth;
+        }
+
+        /// <summary>
+        /// We avoid re-rendering everything while editing if it's possible.
+        /// This method attempts to find the first changed logical line and move the cursor to the right position for the subsequent rendering.
+        /// </summary>
+        private void CalculateWhereAndWhatToRender(bool cursorMovedToInitialPos, ref RenderData renderData, ref RenderedLineInfo current, ref RenderedLineInfo previous)
+        {
+            int bufferWidth = _console.BufferWidth;
+            int bufferHeight = _console.BufferHeight;
+            int cursorX = _console.CursorLeft;
+            int cursorY = _console.CursorTop;
+
+            RenderedLineData[] previousRenderLines = _previousRender.lines;
+            int previousLogicalLine = 0;
+            int previousPhysicalLine = 0;
 
-                    // promptBufferCells is the number of visible characters in the prompt
-                    int promptBufferCells = LengthInBufferCells(promptText);
-                    bool renderErrorPrompt = false;
+            RenderedLineData[] renderLines = renderData.lines;
+            int logicalLine = 0;
+            int physicalLine = 0;
+            int pseudoPhysicalLineOffset = 0;
 
-                    if (_console.CursorLeft >= promptBufferCells)
+            bool hasToWriteAll = true;
+
+            if (cursorY > _initialY && renderLines.Length > 1)
```

Comment:
Then check is an optimization. Actually, we shouldn't even check for `cursorY > _initialY` for this optimization because it would prevent one legit scenario from falling in this `if` code path where `cursorY == _initialY` -- the physical line at `_iniitalY` finishes before the line end; current cursor is at the end of line `_initialY`; the next line is an empty string; the key pressed is <kbd>Delete</kbd>.

For this scenario, we at least don't need to re-write the first logical line, but with the current check, it will fall to the `hasToWriteAll` code path.

Reply:
Updated to remove the `cursorY > _initialY` checking. Also updated the comment following that check.